go/tls/tls.go

```diff
@@ -0,0 +1,682 @@
+			v, err := strconv.ParseUint(part[7:], 10, 64)
+			if err != nil {
+				continue
+			}
+			if info == nil {
+				info = &fieldInfo{}
+			}
+			info.minlen = v
+		case strings.HasPrefix(part, "selector:"):
+			if info == nil {
+				info = &fieldInfo{}
+			}
+			info.selector = part[9:]
+		case strings.HasPrefix(part, "val:"):
+			v, err := strconv.ParseUint(part[4:], 10, 64)
+			if err != nil {
+				continue
+			}
+			if info == nil {
+				info = &fieldInfo{}
+			}
+			info.val = v
+		}
+	}
+	if info != nil {
+		info.name = name
+		if info.selector == "" {
+			var prefix string
+			if info.name != "" {
+				prefix = name + ": "
+			}
+			if info.count < 1 {
+				err = StructuralError{prefix + "field of unknown size in " + str}
+			} else if info.count > 8 {
+				err = StructuralError{prefix + "specified size too large in " + str}
+			} else if info.minlen > info.maxlen {
+				err = StructuralError{prefix + "specified length range inverted in " + str}
+			} else if info.val > 0 {
+				err = StructuralError{prefix + "specified selector value but not field in " + str}
+			}
+		}
+	} else if name != "" {
+		info = &fieldInfo{name: name}
+	}
+	return
+}
+
+// Check that a value fits.
+func (size fieldInfo) check(val uint64, prefix string) error {
+	if val >= (1 << (8 * size.count)) {
+		return StructuralError{fmt.Sprintf("%svalue %d too large for size", prefix, val)}
+	}
+	if size.maxlen != 0 {
+		if val < size.minlen {
+			return StructuralError{fmt.Sprintf("%svalue %d too small for minimum %d", prefix, val, size.minlen)}
+		}
+		if val > size.maxlen {
+			return StructuralError{fmt.Sprintf("%svalue %d too large for maximum %d", prefix, val, size.maxlen)}
+		}
+	}
+	return nil
+}
+
+// readVaruint reads an unsigned big-endian integer of the given size in
+// bytes.
+func readVaruint(data []byte, info *fieldInfo) (result uint64, err error) {
+	var prefix string
+	if info != nil && len(info.name) > 0 {
+		prefix = info.name + ": "
+	}
+	if info == nil || !info.countSet {
+		return 0, StructuralError{prefix + "no field size information available"}
+	}
+	if len(data) < int(info.count) {
+		return 0, SyntaxError{prefix + "truncated variable-length integer"}
+	}
+	for i := uint(0); i < info.count; i++ {
+		result = (result << 8) | uint64(data[i])
+	}
+	if err = info.check(result, prefix); err != nil {
+		return 0, err
+	}
+	return
+}
+
+// parseField is the main parsing function. Given a byte slice and an offset
+// into the data, it will try to parse a suitable ASN.1 value out and store it
+// in the given Value.
+func parseField(v reflect.Value, data []byte, initOffset int, info *fieldInfo) (offset int, err error) {
+	var prefix string
+	if info != nil && len(info.name) > 0 {
+		prefix = info.name + ": "
+	}
+	offset = initOffset
+	rest := data[offset:]
+
+	fieldType := v.Type()
+	// First look for known fixed types.
+	switch fieldType {
+	case uint8Type:
+		if len(rest) < 1 {
+			err = SyntaxError{prefix + "truncated uint8"}
```

Comment:
... and do a direct `return SyntaxError{...}` or equivalent further down for err.